src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1268,15 +1268,43 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         let opaque_defn_ty = tcx.type_of(opaque_def_id);
                         let opaque_defn_ty = opaque_defn_ty.subst(tcx, opaque_decl.substs);
                         let opaque_defn_ty = renumber::renumber_regions(infcx, &opaque_defn_ty);
+                        let concrete_is_opaque = infcx
+                            .resolve_vars_if_possible(&opaque_decl.concrete_ty).is_impl_trait();
+
                         debug!(
-                            "eq_opaque_type_and_type: concrete_ty={:?}={:?} opaque_defn_ty={:?}",
+                            "eq_opaque_type_and_type: concrete_ty={:?}={:?} opaque_defn_ty={:?} \
+                            concrete_is_opaque={}",
                             opaque_decl.concrete_ty,
                             infcx.resolve_vars_if_possible(&opaque_decl.concrete_ty),
-                            opaque_defn_ty
+                            opaque_defn_ty,
+                            concrete_is_opaque
                         );
-                        obligations.add(infcx
-                            .at(&ObligationCause::dummy(), param_env)
-                            .eq(opaque_decl.concrete_ty, opaque_defn_ty)?);
+
+                        // concrete_is_opaque is `true` when we're using an existential
+                        // type without 'revealing' it. For example, code like this:
+                        //
+                        // existential type Foo: Debug;
+                        // fn foo1() -> Foo { ... }
+                        // fn foo2() -> Foo { foo1() }
+                        //
+                        // In `foo2`, we're not revealing the type of `Foo` - we're
+                        // just treating it as the opaque type.
+                        //
+                        // When this occurs, we do *not* want to try to equate
+                        // the concrete type with the underlying defining type
+                        // of the existential type - this will always fail, since
+                        // the defining type of an existential type is always
+                        // some other type (e.g. not itself)
+                        // Essentially, none of the normal obligations apply here -
+                        // we're just passing around some unknown opaque type,
+                        // without actually looking at the underlying type it
+                        // gets 'revealed' into
+
+                        if !concrete_is_opaque {
```

Comment:
if the concrete type is opaque, don't we need an additional check to figure out if the opaque types are equal? Essentially what I'm asking is whether

```rust
pub trait MyTrait {}

#[derive(Debug)]
pub struct MyStruct {
  v: u64
}

impl MyTrait for MyStruct {}

mod foo {
    pub fn bla() -> TE {
        return crate::MyStruct {v:1}
    }
    
    
    existential type TE: crate::MyTrait;
}

mod bar {
    pub fn bla2() -> TE2 {
        crate::foo::bla()
    }
    
    pub existential type TE2: crate::MyTrait;
}

pub fn bla2() -> bar::TE2 {
    crate::foo::bla()
}
```

still errors after your PR

Reply:
@oli-obk See my reply to @nikomatsakis above

Reply:
The code you linked is to the master branch (or at least before your PR) right here. So we aren't always checking that as you wrote.

Reply:
Yes - that line is [unchanged]9https://github.com/Aaron1011/rust/blob/2f4196205336df8550a4bfb3045d3d1c350f02bf/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L1263-L1265) in my PR - that is, we still equate the opaque types.

Reply:
:man_facepalming: 	sorry, These two obligations looked so similar, I confused the two.